### PR TITLE
Add bluetooth widget

### PR DESCRIPTION
**Pull request by Layerex**
Add a widget to display currently connected and paired bluetooth devices.

### CODE REVIEW

File: examples/dwm/bluetooth.lua [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+-- [paired]        boolean true
+-- [trusted]       boolean true
+function get_device_info(mac_address)
+    if mac_address == nil then
+        mac_address = ""
+    end
+    local device_info = {}
+    local handle = io.popen(string.format("bluetoothctl info %s", mac_address))
+    for line in handle:lines() do
+        local key, value = string.match(line, "(%w+): ([%w%s]+)")
```
Comment (shdown) [inline examples/dwm/bluetooth.lua:58]: Use `.*` rather than `[%w%s]+`.

File: examples/dwm/bluetooth.lua [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        end
+    end
+    handle:close()
+    return device_info
+end
+
+function reprint_devices()
+    local s = ""
+    for mac_address, device in pairs(devices) do
+        s = s .. string.format("%s(%s)", device["name"], mac_address) .. separator
```
Comment (shdown) [inline examples/dwm/bluetooth.lua:80]: This is O(N^2). Instead, please do something like
```lua
local t = {}
for mac_address, device in pairs(devices) do
    table.insert(t, string.format('%s(%s)', device.name, mac_address))
end
return table.concat(t, separator)
```

File: examples/dwm/bluetooth.lua [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    end
+    return string.sub(s, 1, #s - 1) -- remove last separator
```
Comment (shdown) [inline examples/dwm/bluetooth.lua:82]: (Note also that this assumes `#separator == 1`.)

File: examples/dwm/bluetooth.lua [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                end
+            end
+            device_info[key] = value
+        end
+    end
+    handle:close()
+    return device_info
+end
+
+function reprint_devices()
```
Comment (shdown) [inline examples/dwm/bluetooth.lua:77]: This function operates on a global variable `devices`, which is defined later. Please move the assignment.

**Review by shdown — CHANGES_REQUESTED** [review]
Thanks for contributing! This will be merged after the issues I mentioned are fixed.

File: examples/dwm/bluetooth.lua [hunk 5]
```diff
@@ -0,0 +1,141 @@
+-- A widget to display currently connected and paired bluetooth devices.
+-- To change output format modify reprint_devices function.
+
+separator = " "
+
+-- Object paths look like /org/bluez/hci0/dev_XX_XX_XX_XX_XX_XX/somethingsomething
+function get_device_mac_address(device_object_path)
+    return device_object_path:gsub("/.*/dev_", ""):gsub("/.*", ""):gsub("_", ":")
```
Comment (shdown) [inline examples/dwm/bluetooth.lua:8]: What if `/somethingsometing` contains `/dev_`?

I would rewrite it as follows:
```lua
function get_device_mac_address(obj_path)
    local s = obj_path:match('/dev_([%x_]+)')
    return s and s:gsub('_', ':')
end
```

File: examples/i3/bluetooth.lua [hunk 6]
```diff
@@ -0,0 +1,141 @@
+-- A widget to display currently connected and paired bluetooth devices.
+-- To change output format modify reprint_devices function.
+
+separator = " "
+
+-- Object paths look like /org/bluez/hci0/dev_XX_XX_XX_XX_XX_XX/somethingsomething
+function get_device_mac_address(device_object_path)
```
Comment (shdown) [inline examples/i3/bluetooth.lua:7]: Same here.

File: examples/i3/bluetooth.lua [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+-- [paired]        boolean true
+-- [trusted]       boolean true
+function get_device_info(mac_address)
+    if mac_address == nil then
+        mac_address = ""
+    end
+    local device_info = {}
+    local handle = io.popen(string.format("bluetoothctl info %s", mac_address))
+    for line in handle:lines() do
+        local key, value = string.match(line, "(%w+): ([%w%s]+)")
```
Comment (shdown) [inline examples/i3/bluetooth.lua:58]: Same issue as in the `dwm` counterpart.

File: examples/i3/bluetooth.lua [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+function reprint_devices()
+    local s = ""
+    for mac_address, device in pairs(devices) do
+        s = s .. string.format("%s(%s)", device["name"], mac_address) .. separator
+    end
+    return string.sub(s, 1, #s - 1) -- remove last separator
+end
+
+devices = {}
```
Comment (shdown) [inline examples/i3/bluetooth.lua:85]: Same issue as in the `dwm` counterpart.

File: examples/i3/bluetooth.lua [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        end
+    end
+    handle:close()
+    return device_info
+end
+
+function reprint_devices()
+    local s = ""
+    for mac_address, device in pairs(devices) do
+        s = s .. string.format("%s(%s)", device["name"], mac_address) .. separator
```
Comment (shdown) [inline examples/i3/bluetooth.lua:80]: Same issue as in the `dwm` counterpart.

**Review by shdown — CHANGES_REQUESTED** [review]

File: examples/dwm/bluetooth.lua (hunk 5 above)
Comment (Layerex) [reply examples/dwm/bluetooth.lua:8]: Function won't encounter such string, probably, at least bluez documentation does not specify this. screen (https://git.kernel.org/pub/scm/bluetooth/bluez.git/tree/doc)
Also object path returned by dbus plugin shouldn't contain `/somethingsomething` at all; stripping it is a leftover from earlier versions of a widget, where I tried to use different signals to detect devices connecting and disconnecting.

**Pull request merged**